source/common/ur_pool_handle.hpp

```diff
@@ -0,0 +1,152 @@
+                for (auto &config : DisjointPoolConfigs.Configs) {
+                    config.MaxPoolableSize = Limits->maxPoolableSize;
+                    config.SlabMinSize = Limits->minDriverAllocSize;
+                }
+                break;
+            }
+            default: {
+                throw umf::UsmAllocationException(
+                    UR_RESULT_ERROR_INVALID_ARGUMENT);
+            }
+            }
+            pNext = BaseDesc->pNext;
+        }
+
+        ur_result_t Ret;
+        std::tie(Ret, PoolManager) =
+            usm::pool_manager<usm::pool_descriptor>::create();
+        if (Ret) {
+            throw umf::UsmAllocationException(Ret);
+        }
+
+        auto UrUSMPool = reinterpret_cast<ur_usm_pool_handle_t>(this);
+
+        // We set up our pool handle with default pools for each of the three
+        // allocation types
+        std::vector<usm::pool_descriptor> Descs;
+        std::tie(Ret, Descs) =
+            usm::pool_descriptor::createDefaults(UrUSMPool, Context);
+
+        for (auto &Desc : Descs) {
+            umf::pool_unique_handle_t Pool = nullptr;
+            auto PoolType = usm::urTypeToDisjointPoolType(Desc.type);
+            std::tie(Ret, Pool) =
+                createUMFPoolForDesc<usm::DisjointPool, HostProvider,
+                                     DeviceProvider, SharedProvider>(
+                    Desc, this->DisjointPoolConfigs.Configs[PoolType]);
+            if (Ret) {
+                throw umf::UsmAllocationException(Ret);
+            }
+
+            PoolManager.addPool(Desc, Pool);
+        }
+    }
+
+    std::atomic_uint32_t RefCount = 1;
+    ur_context_handle_t Context = nullptr;
+    usm::DisjointPoolAllConfigs DisjointPoolConfigs =
+        usm::DisjointPoolAllConfigs();
+    usm::pool_manager<usm::pool_descriptor> PoolManager;
+    ur_usm_pool_flags_t PoolFlags;
+
+    uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+    uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+    uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+    bool hasUMFPool(umf_memory_pool_t *umf_pool) {
+        return PoolManager.hasPool(umf_pool);
+    }
+
+    std::optional<umf_memory_pool_handle_t>
+    getOrCreatePool(usm::pool_descriptor &desc) {
+        auto foundPool = PoolManager.getPool(desc);
+        if (foundPool.has_value()) {
+            return foundPool.value();
+        }
+
+        umf::pool_unique_handle_t newPool;
+        ur_result_t Ret = UR_RESULT_SUCCESS;
+        auto PoolType = usm::urTypeToDisjointPoolType(desc.type);
+        std::tie(Ret, newPool) =
+            createUMFPoolForDesc<usm::DisjointPool, HostProvider,
+                                 DeviceProvider, SharedProvider>(
+                desc, this->DisjointPoolConfigs.Configs[PoolType]);
+        if (Ret) {
+            throw umf::UsmAllocationException(Ret);
+        }
+
+        PoolManager.addPool(desc, newPool);
```

Comment:
nit: I think it would be less error-prone to make addPool accept uniquq_ptr by value so that the caller would need to std::move() the argument.